livekit-android-sdk/src/main/java/io/livekit/android/room/SignalClient.kt

```diff
@@ -548,26 +562,68 @@ constructor(
             LivekitRtc.SignalResponse.MessageCase.TRACK_UNPUBLISHED -> {
                 listener?.onLocalTrackUnpublished(response.trackUnpublished)
             }
+            LivekitRtc.SignalResponse.MessageCase.PONG -> {
+                resetPingTimeout()
+            }
             LivekitRtc.SignalResponse.MessageCase.MESSAGE_NOT_SET,
             null -> {
                 LKLog.v { "empty messageCase!" }
             }
-        }.safe()
+        }
+    }
+
+    private fun startPingJob() {
+        if (pingJob == null && pingIntervalDurationMillis != 0L) {
+            pingJob = coroutineScope.launch {
+                while (true) {
+                    delay(pingIntervalDurationMillis)
+
+                    val pingTimestamp = Date().time
+                    val pingRequest = LivekitRtc.SignalRequest.newBuilder()
+                        .setPing(pingTimestamp)
+                        .build()
+                    LKLog.v { "Sending ping: $pingTimestamp" }
+                    sendRequest(pingRequest)
+                    startPingTimeout(pingTimestamp)
+                }
+            }
+        }
+    }
+
+    private fun startPingTimeout(timestamp: Long) {
+        if (pongJob != null) {
+            return
+        }
+        pongJob = coroutineScope.launch {
+            delay(pingTimeoutDurationMillis)
+            LKLog.d { "Ping timeout reached for ping sent at $timestamp." }
+            currentWs?.close(CLOSE_REASON_PING_TIMEOUT, "Ping timeout")
```

Comment:
would this attempt a reconnect automatically?

Reply:
Yeah, it kicks off a reconnection (RTCEngine observes the signal client for websocket closure/failure).